### PR TITLE
Overhaul of torchseg.losses

### PR DESCRIPTION
**Pull request by notprime**
Overhaul of torchseg.losses, starting from DiceLoss.

General changes applied to DiceLoss (to be applied later to all other losses):

- switched from `assert` to `warnings` and `raise`;
- eliminated the need of specifying `binary`, `multiclass` and `multilabels` modes when instantiating a loss function, by changing masks shape: now the shape must be [B, C, H, W] for multiclass and multilabels cases, [B, 1, H, W] for binary case. [B, 1, H, W] can be accepted also for multiclass case as long as `mask_to_one_hot = True`;
- introduction of proper reduction techniques (`MEAN`, `SUM`, `NONE`) and definition of a `LossReduction` class to easily save the techniques;

Specific changes applied to DiceLoss:

- added the argument `power` to be applied to the denominator of the DiceCoefficient (still need to add this to the documentation of the class)

### CODE REVIEW

**Comment by notprime** [comment]
@isaaccorley let me know what do you think about the "new" DiceLoss, it seems cleaner and more efficient without the burden of specyfying `binary`, `multilabel` and `multiclass` modes.

I'm not particularly sure about line 135 and 136 in the `forward` of DiceLoss, it does not make much sense to me for the two reasons I've written in the comment.

File: torchseg/losses/_functional.py [hunk 1]
```diff
@@ -151,32 +151,37 @@ def soft_jaccard_score(
...
 
 def soft_dice_score(
     output: torch.Tensor,
     target: torch.Tensor,
+    power: float =  1.0,
     smooth: float = 0.0,
     eps: float = 1e-7,
-    dims=None,
+    dims = None,
 ) -> torch.Tensor:
```
Comment (isaaccorley) [inline torchseg/losses/_functional.py:172]: Can you add a docstring to this?

File: torchseg/losses/_functional.py [hunk 2]
```diff
@@ -151,32 +151,37 @@ def soft_jaccard_score(
...
+        target_pow = torch.sum(target ** power, dim = dims)
+        cardinality = output_pow + target_pow
     else:
         intersection = torch.sum(output * target)
-        cardinality = torch.sum(output + target)
-    dice_score = (2.0 * intersection + smooth) / (cardinality + smooth).clamp_min(eps)
+        output_pow = torch.sum(output ** power)
+        target_pow = torch.sum(target ** power)
+        cardinality = output_pow + target_pow
+    dice_score = (2.0 * intersection + smooth) / (cardinality + eps)
```
Comment (isaaccorley) [inline torchseg/losses/_functional.py:184]: Do we not need to add smooth to cardinality?

File: torchseg/losses/dice.py [hunk 3]
```diff
@@ -1,140 +1,147 @@
...
-            if self.ignore_index is not None:
-                mask = y_true != self.ignore_index
-                y_pred = y_pred * mask
-                y_true = y_true * mask
-
-        if self.mode == MULTICLASS_MODE:
-            y_true = y_true.view(bs, -1)
-            y_pred = y_pred.view(bs, num_classes, -1)
-
-            if self.ignore_index is not None:
```
Comment (isaaccorley) [inline torchseg/losses/dice.py:91]: We are removing the ignore_index filtering but I don't see it added below in the new code

File: torchseg/losses/dice.py [hunk 4]
```diff
@@ -1,140 +1,147 @@
...
-        )
+        if self.mask_to_one_hot:
+            # Convert y_true to one_hot representation to compute DiceLoss
+            if num_classes == 1:
+                warnings.warn("Single channel prediction, 'mask_to_one_hot = True' ignored.")
+            else:
+                # maybe there is a better way to handle this?
+                permute_dims = tuple(dim - 1 for dim in spatial_dims)
+                y_true = F.one_hot(y_true, num_classes).squeeze(dim = 1) # N, 1, H, W, ... ---> N, H, W, ..., C
+                y_true = y_true.permute(0, -1, *permute_dims) # N, 1, H, W, ..., C ---> N, C, H, W, ...
```
Comment (isaaccorley) [inline torchseg/losses/dice.py:99]: You can use einops for this. Its cleaner

File: torchseg/losses/dice.py [hunk 5]
```diff
@@ -1,140 +1,147 @@
...
+            loss = loss[:, self.classes, :]
 
-    def aggregate_loss(self, loss):
-        return loss.mean()
+        if self.reduction == LossReduction.MEAN:
+            loss = torch.mean(loss)
+        elif self.reduction == LossReduction.SUM:
+            loss = torch.sum(loss)
+        elif self.reduction == LossReduction.NONE:
+            broadcast_shape = list(loss.shape[0:2]) + [1] * (len(y_true.shape) - 2)
```
Comment (isaaccorley) [inline torchseg/losses/dice.py:144]: What does this do?

**Review by isaaccorley — CHANGES_REQUESTED** [review]
This looks really clean. I like it. Just a few comments. We should also add tests

File: torchseg/losses/_functional.py (hunk 2 above)
Comment (notprime) [reply torchseg/losses/_functional.py:184]: eps and smooth basically do the same thing (smooth avoid a value of 0, eps avoid nan). Supposing that smooth should be very small (same order as eps), we could just use one of the two, but we can also treat them separately, let me know

i think leaving them both gives more customization

File: torchseg/losses/dice.py (hunk 5 above)
Comment (notprime) [reply torchseg/losses/dice.py:144]: just giving the users the ability to choose a preferred reduction instead of the classic `mean`. The latter basically returns a DiceScore per channel. So, if y_true has shape [B, C, H, W], selecting `LossReduction.NONE` returns a loss with shape [B, C, 1, 1].
